File: host/src/models/digest.ts
/** Streaming sha256 of a cached file, so a multi-gigabyte model never lands in memory. */
import { createHash } from 'node:crypto'
import { createReadStream } from 'node:fs'

export interface Digest {
  readonly sha256: string
  readonly sizeBytes: number
}

export async function sha256Of(path: string): Promise<Digest> {
  const hash = createHash('sha256')
  let sizeBytes = 0
  for await (const chunk of createReadStream(path, { highWaterMark: 1 << 20 })) {
    const bytes = chunk as Buffer
    sizeBytes += bytes.length
    hash.update(bytes)
  }
  return { sha256: hash.digest('hex'), sizeBytes }
}

File: models/src/digest.rs
//! Streaming sha256 of a cached file, so a multi-gigabyte model never lands in memory.

use sha2::{Digest, Sha256};
use std::fs::File;
use std::io::Read;
use std::path::Path;

pub fn sha256_of(path: &Path) -> std::io::Result<(String, u64)> {
    let mut file = File::open(path)?;
    let mut hasher = Sha256::new();
    let mut buf = vec![0u8; 1 << 20];
    let mut size = 0u64;
    loop {
        let read = file.read(&mut buf)?;
        if read == 0 {
            break;
        }
        size += read as u64;
        hasher.update(&buf[..read]);
    }
    let hex = hasher.finalize().iter().map(|b| format!("{b:02x}")).collect();
    Ok((hex, size))
}
